azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/AbstractAzureMojo.java

```diff
@@ -272,74 +274,97 @@ public int getHttpProxyPort() {
             } else {
-                initAuth();
-                azure = getAzureClientByAuthType();
+                azure = createAzureClient();
             }
             if (azure == null) {
                 getTelemetryProxy().trackEvent(INIT_FAILURE);
                 throw new AzureAuthFailureException(AZURE_INIT_FAIL);
             }
             printCurrentSubscription(azure);
-            getTelemetryProxy().addDefaultProperty(AUTH_TYPE, authType);
+            getTelemetryProxy().addDefaultProperty(AUTH_TYPE, getAuthType());
             getTelemetryProxy().addDefaultProperty(AUTH_METHOD, getAuthMethod());
             // Repopulate subscriptionId in case it is not configured.
             getTelemetryProxy().addDefaultProperty(SUBSCRIPTION_ID_KEY, azure.subscriptionId());
-
         }
         return azure;
     }
 
-    protected Azure getAzureClientByAuthType() throws AzureAuthFailureException {
+    protected String getAuthType() {
+        return StringUtils.firstNonBlank(auth == null ? null : auth.getAuthType(), authType);
+    }
+
+    protected Subscription selectSubscription(Azure az, TextIO textIO, Subscription[]subscriptions) throws AzureExecutionException {
+        if (subscriptions.length == 0) {
+            throw new AzureExecutionException("Cannot find any subscriptions in current account.");
+        }
+        if (subscriptions.length == 1) {
+            Log.info(String.format("There is only one subscription '%s' in your account, will use it automatically.",
+                    TextUtils.blue(SubscriptionOption.getSubscriptionName(subscriptions[0]))));
+            return subscriptions[0];
+        }
+        final String defaultId = Optional.ofNullable(az.getCurrentSubscription()).map(Subscription::subscriptionId).orElse(null);
+
+        final List<SubscriptionOption> wrapSubs = Arrays.stream(subscriptions).map(t -> new SubscriptionOption(t))
+                .sorted()
+                .collect(Collectors.toList());
+        final SubscriptionOption defaultValue = wrapSubs.stream()
+                .filter(t -> StringUtils.equalsIgnoreCase(t.getSubscriptionId(), defaultId)).findFirst().orElse(null);
+
+        final SubscriptionOption subscriptionOptionSelected = new CustomTextIoStringListReader<SubscriptionOption>(() -> textIO.getTextTerminal(), null)
+                .withCustomPrompt(String.format("Please choose a subscription%s: ",
+                        highlightDefaultValue(defaultValue == null ? null : defaultValue.getSubscriptionName())))
+                .withNumberedPossibleValues(wrapSubs).withDefaultValue(defaultValue).read("Available subscriptions:");
+        if (subscriptionOptionSelected == null) {
+            throw new AzureExecutionException("You must select a subscription.");
+        }
+        return subscriptionOptionSelected.getSubscription();
+    }
+
+    protected Azure createAzureClient() throws AzureAuthFailureException, AzureExecutionException {
```

Comment:
this method has 2 side-effects: `this.auth` & `this.subscriptionId` is changed. 
is this good?